File: internal/study/choice.go
// Package study holds shared study-mode logic used by multiple screens:
// multiple-choice option generation and study-streak bookkeeping.
package study

import "math/rand"

// Options returns up to n answer options containing correct plus distinct
// distractors drawn from pool, along with the index of the correct option. The
// ordering is randomized using rng so the result is deterministic in tests.
func Options(rng *rand.Rand, correct string, pool []string, n int) ([]string, int) {
	seen := map[string]bool{correct: true}
	distractors := make([]string, 0, len(pool))
	for _, p := range pool {
		if !seen[p] {
			seen[p] = true
			distractors = append(distractors, p)
		}
	}
	rng.Shuffle(len(distractors), func(i, j int) {
		distractors[i], distractors[j] = distractors[j], distractors[i]
	})

	want := n - 1
	if want > len(distractors) {
		want = len(distractors)
	}
	options := make([]string, 0, want+1)
	options = append(options, correct)
	options = append(options, distractors[:want]...)
	rng.Shuffle(len(options), func(i, j int) {
		options[i], options[j] = options[j], options[i]
	})

	correctIdx := 0
	for i, o := range options {
		if o == correct {
			correctIdx = i
			break
		}
	}
	return options, correctIdx
}

File: core/src/study/choice.rs
use std::collections::HashSet;

use rand::seq::SliceRandom;
use rand::Rng;

/// Returns up to `n` answer options containing `correct` plus distinct
/// distractors drawn from `pool`, along with the index of the correct option.
/// The ordering is randomized using `rng` so the result is deterministic in
/// tests.
pub fn options<R: Rng + ?Sized>(
    rng: &mut R,
    correct: &str,
    pool: &[String],
    n: usize,
) -> (Vec<String>, usize) {
    let mut seen: HashSet<&str> = HashSet::new();
    seen.insert(correct);
    let mut distractors: Vec<String> = Vec::with_capacity(pool.len());
    for p in pool {
        if seen.insert(p.as_str()) {
            distractors.push(p.clone());
        }
    }
    distractors.shuffle(rng);

    let want = n.saturating_sub(1).min(distractors.len());
    let mut opts = Vec::with_capacity(want + 1);
    opts.push(correct.to_string());
    opts.extend_from_slice(&distractors[..want]);
    opts.shuffle(rng);

    let correct_idx = opts.iter().position(|o| o == correct).unwrap_or(0);
    (opts, correct_idx)
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::rngs::StdRng;
    use rand::SeedableRng;

    fn pool(items: &[&str]) -> Vec<String> {
        items.iter().map(|s| s.to_string()).collect()
    }

    #[test]
    fn includes_correct_and_indexes_it() {
        let mut rng = StdRng::seed_from_u64(1);
        let (opts, idx) = options(&mut rng, "a", &pool(&["b", "c", "d", "e"]), 4);
        assert_eq!(opts.len(), 4);
        assert_eq!(opts[idx], "a");
        assert!(opts.contains(&"a".to_string()));
    }

    #[test]
    fn dedups_pool_and_caps_gracefully() {
        let mut rng = StdRng::seed_from_u64(2);
        // Pool has duplicates and the correct answer; only 2 distinct distractors.
        let (opts, idx) = options(&mut rng, "a", &pool(&["a", "b", "b", "c"]), 4);
        assert_eq!(opts.len(), 3); // correct + 2 distinct distractors
        assert_eq!(opts[idx], "a");
    }
}
